mllib/src/main/scala/org/apache/spark/mllib/clustering/PIClustering.scala

```diff
@@ -0,0 +1,433 @@
+          minAffinity: Double = DefaultMinAffinity,
+          nRuns: Int = DefaultKMeansRuns)
+  : (Seq[(Int, Vector)], Seq[((VertexId, Vector), Int)]) = {
+    val vidsRdd = sc.parallelize(points.map(_._1).sorted)
+    val nVertices = points.length
+
+    val (wRdd, rowSums) = createNormalizedAffinityMatrix(sc, points, sigma)
+    val initialVt = createInitialVector(sc, points.map(_._1), rowSums)
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Vt(0)=${
+        printVector(new BDV(initialVt.map {
+          _._2
+        }.toArray))
+      }")
+    }
+    val edgesRdd = createSparseEdgesRdd(sc, wRdd, minAffinity)
+    val G = createGraphFromEdges(sc, edgesRdd, points.size, Some(initialVt))
+    if (logger.isDebugEnabled) {
+      logger.debug(printMatrixFromEdges(G.edges))
+    }
+    val (gUpdated, lambda, vt) = getPrincipalEigen(sc, G, nIterations)
+    // TODO: avoid local collect and then sc.parallelize.
+    val localVt = vt.collect.sortBy(_._1)
+    val vectRdd = sc.parallelize(localVt.map(v => (v._1, Vectors.dense(v._2))))
+    vectRdd.cache()
+    val model = KMeans.train(vectRdd.map {
+      _._2
+    }, nClusters, nRuns)
+    vectRdd.unpersist()
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Eigenvalue = $lambda EigenVector: ${localVt.mkString(",")}")
+    }
+    val estimates = vectRdd.zip(model.predict(vectRdd.map(_._2)))
+    if (logger.isDebugEnabled) {
+      logger.debug(s"lambda=$lambda  eigen=${localVt.mkString(",")}")
+    }
+    val ccs = (0 until model.clusterCenters.length).zip(model.clusterCenters)
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Kmeans model cluster centers: ${ccs.mkString(",")}")
+    }
+    val estCollected = estimates.collect.sortBy(_._1._1)
+    if (logger.isDebugEnabled) {
+      val clusters = estCollected.map(_._2)
+      val counts = estCollected.groupBy(_._2).mapValues {
+        _.length
+      }
+      logger.debug(s"Cluster counts: Counts: ${counts.mkString(",")}"
+        + s"\nCluster Estimates: ${estCollected.mkString(",")}")
+    }
+    (ccs, estCollected)
+  }
+
+  /**
+   * Read Points from an input file in the following format:
+   * Vertex1Id Coord11 Coord12 CoordX13 .. Coord1D
+   * Vertex2Id Coord21 Coord22 CoordX23 .. Coord2D
+   * ..
+   * VertexNId CoordN1 CoordN2 CoordN23 .. CoordND
+   *
+   * Where N is the number of observations, each a D-dimension point
+   *
+   * E.g.
+   *
+   * 19	1.8035177495	0.7460582552	0.2361611395	-0.8645567427	-0.8613062
+   * 10	0.5534111111	1.0456386879	1.7045663273	0.7281759816	1.0807487792
+   * 911	1.200749626	1.8962364439	2.5117192131	-0.4034737281	-0.9069696484
+   *
+   * Which represents three 5-dimensional input Points with VertexIds 19,10, and 911
+   * @param verticesFile Local filesystem path to the Points input file
+   * @return Set of Vertices in format appropriate for consumption by the PIC algorithm
+   */
+  def readVerticesfromFile(verticesFile: String): Points = {
```

Comment:
Let's not handle I/O here. We can have an example code under `examples/` and load files there.